test/jasmine/tests/select_test.js

```diff
@@ -545,108 +584,70 @@ describe('select box and lasso', function() {
-            var deselectCnt = 0;
-
-            gd.once('plotly_selecting', function() {
-                assertSelectionNodes(1, 2);
-                selectingCnt++;
-            });
-
-            gd.once('plotly_selected', function(d) {
-                assertSelectionNodes(0, 2);
-                selectedCnt++;
-                eventData = d;
-            });
-
-            gd.once('plotly_deselect', function() {
-                deselectCnt++;
-                assertSelectionNodes(0, 0);
-            });
+        function checkEventCounts(selecting, selected, deselect, msg) {
+            expect(selectingCnt).toBe(selecting, 'plotly_selecting call count: ' + msg);
+            expect(selectedCnt).toBe(selected, 'plotly_selected call count: ' + msg);
+            expect(deselectCnt).toBe(deselect, 'plotly_deselect call count: ' + msg);
+        }
 
+        Plotly.plot(gd, fig).then(function() {
+            resetEvents(gd);
             drag([[370, 120], [500, 200]]);
+            return selectedPromise;
+        })
+        .then(function() {
             assertPoints([{lon: 30, lat: 30}]);
             assertRanges([[21.99, 34.55], [38.14, 25.98]]);
+            checkEventCounts(1, 1, 0, 'select');
 
-            return doubleClick(250, 200).then(function() {
-                expect(selectingCnt).toBe(1, 'plotly_selecting call count');
-                expect(selectedCnt).toBe(1, 'plotly_selected call count');
-                expect(deselectCnt).toBe(1, 'plotly_deselect call count');
-            });
+            return doubleClick(250, 200);
+        })
+        .then(deselectPromise)
+        .then(function() {
+            // deselect gives an extra plotly_selected event on the first click
+            // event data is undefined
+            checkEventCounts(1, 2, 1, 'deselect');
         })
         .then(function() {
             return Plotly.relayout(gd, 'dragmode', 'lasso');
         })
         .then(function() {
-            var selectingCnt = 0;
-            var selectedCnt = 0;
-            var deselectCnt = 0;
-
-            gd.once('plotly_selecting', function() {
-                assertSelectionNodes(1, 2);
-                selectingCnt++;
-            });
-
-            gd.once('plotly_selected', function(d) {
-                assertSelectionNodes(0, 2);
-                selectedCnt++;
-                eventData = d;
-            });
-
-            gd.once('plotly_deselect', function() {
-                deselectCnt++;
-                assertSelectionNodes(0, 0);
-            });
-
+            resetEvents(gd);
             drag([[300, 200], [300, 300], [400, 300], [400, 200], [300, 200]]);
+            return selectedPromise;
+        })
+        .then(function() {
             assertPoints([{lon: 20, lat: 20}]);
             assertLassoPoints([
                 [13.28, 25.97], [13.28, 14.33], [25.71, 14.33], [25.71, 25.97], [13.28, 25.97]
             ]);
+            checkEventCounts(4, 1, 0, 'lasso');
```

Comment:
4 `plotly_selecting` events is real - because there are 5 points in the lasso path.